tests/test_io_sensors.py

```diff
@@ -0,0 +1,171 @@
+from .pigpio_mocks import *
+from vent.io.devices import ADS1015, ADS1115, PigpioConnection
+from vent.io.devices.sensors import Sensor, AnalogSensor, SFM3200, SimSensor
+
+
+def test_sensor():
+    """__________________________________________________________________________________________________________TEST #1
+    Tests errors are thrown when you try to do stuff with the abstract base class
+    """
+    with pytest.raises(TypeError):
+        s = Sensor()
+
+
+def test_sim_sensor(patch_pigpio_i2c, mock_i2c_hardware):
+    """__________________________________________________________________________________________________________TEST #2
+    Tests the proper functioning of a simulated sensor. Not much to this one, really. Also checks that age sends back -1
+        when update() has not been called
+    """
+    sim_sensor = SimSensor()
+    assert sim_sensor.age() == -1
+    n_iter = random.randint(1, 1000)
+    for i in range(n_iter):
+
+        result = sim_sensor.get()
+        assert sim_sensor.low <= result <= sim_sensor.high
+    sim_sensor.update()
+    assert sim_sensor.age() > 0
+    """__________________________________________________________________________________________________________
+    """
+
+
+@pytest.mark.parametrize("ads1x15", [ADS1115, ADS1015])
+@pytest.mark.parametrize("seed", [secrets.token_bytes(8) for _ in range(64)])
+@pytest.mark.parametrize("pga", random.sample(ADS1115._CONFIG_VALUES[2], 4))
+@pytest.mark.parametrize("mode", ['SINGLE', 'CONTINUOUS'])
+def test_analog_sensor_single_read(patch_pigpio_i2c, mock_i2c_hardware, ads1x15, seed, pga, mode):
+    """__________________________________________________________________________________________________________TEST #3
+    Tests the proper functioning of an AnalogSensor with random, but plausible, calibration.
+
+    - Performs a sequence of observations and verifies that each matches expectations and that the age of each reading
+        properly reflects called updates
+    """
+    random.seed(seed)
+    kwargs = {
+        "MUX": random.choice(ads1x15._CONFIG_VALUES[1]),
+        "PGA": pga,
+        "MODE": mode,
+        "DR": random.choice(ads1x15._CONFIG_VALUES[4]),
+        'offset_voltage': round(random.uniform(-0.35, 0.35)*pga, 2),
+        'output_span': round(random.uniform(.75, 1.0)*pga, 2),
+        'conversion_factor': round(random.choice([-1, 1])*random.uniform(1, 10), 2)
+    }
+    conversion_factor = kwargs['conversion_factor']
+    if random.getrandbits(1):
+        # occasionally drop a kwarg or two to force a default check_and_set
```

Comment:
what does this do?

Reply:
I broke this test out a bit. Might be slightly more clear now